core/storage/question/gae_models.py

```diff
@@ -261,7 +261,84 @@ def get_question_skill_links_by_skill_ids(
+    def get_sorted_question_skill_links_equidistributed_by_skill(
+            cls, total_question_count, skill_ids, degrees_of_mastery):
+        """Fetches the list of constant number of QuestionSkillLinkModels
+        linked to the skills, sorted by the difference between skill
+        difficulty and skill mastery.
+
+        Args:
+            total_question_count: int. The number of questions expected.
+            skill_ids: list(str). The ids of skills for which the linked
+                question ids are to be retrieved.
+            degrees_of_mastery: dict(str, float). The keys are the requested
+                skill IDs. The values are the corresponding mastery degree of
+                the user.
+
+        Returns:
+            list(QuestionSkillLinkModel). A list of QuestionSkillLinkModels
+                corresponding to given skill_ids, with
+                total_question_count/len(skill_ids) number of questions for
+                each skill. If not evenly divisible, it will be rounded up.
+                If not enough questions for a skill, just return all questions
+                it links to. The order of questions will follow the order of
+                given skill ids, and the order of questions for the same skill
+                follows the difference between skill difficulty and mastery.
+        """
+        question_count_per_skill = int(
+            math.ceil(float(total_question_count) / float(len(skill_ids))))
+
+        question_skill_link_models = []
+
+        for skill_id in skill_ids:
+            degree_of_mastery = degrees_of_mastery[skill_id]
+            if not degree_of_mastery:
+                degree_of_mastery = 0.0
+
+            query = cls.query(cls.skill_id == skill_id)
+
+            # Fetch QuestionSkillLinkModels with difficulty larger than
+            # mastery and sort them by increasing difficulty.
+            harder_questions_query = query.filter(
+                cls.skill_difficulty >= degree_of_mastery)
+            harder_questions_query = harder_questions_query.order(
+                cls.skill_difficulty)
+            new_question_skill_link_models = (
+                harder_questions_query.fetch(question_count_per_skill))
+
+            # Fetch QuestionSkillLinkModels with difficulty smaller than
+            # mastery and sort them by decreasing difficulty.
+            easier_questions_query = query.filter(
+                cls.skill_difficulty < degree_of_mastery)
+            easier_questions_query = easier_questions_query.order(
+                -cls.skill_difficulty)
+            new_question_skill_link_models.extend(
+                easier_questions_query.fetch(question_count_per_skill))
+
+            # Deduplicate if the same question is linked to multiple skills.
+            # Note: This deduplication might make the number of returning
+            # questions too low.
+            existed_question_ids = [
```

Comment:
@seanlip Since you cannot use multiple filters with different params in a single query, this is what I had suggested to @sophiewu6 , if you have any other ideas please let us know!

Reply:
Seems reasonable, but:
- maybe don't keep recomputing the list of model.question_id each time, instead create a set that gets updated at the end of each iteration
- should we fetch a few more questions than needed to reduce the chance we run into the "too few questions" issue?

Reply:
For the second suggestion, we are already doing that..we are actually fetching double the number than requested anyway (because of the two fetches we are doing above). So the odds of running into fewer questions should be low

Reply:
Ah yes, I missed that. @sophiewu6 I would actually suggest adding that observation as an explicit comment in the code.

Reply:
Done.